ui/ducks/send/send.js

```diff
@@ -0,0 +1,1325 @@
+      from: '',
+      data: undefined,
+      value: '0x0',
+      gas: '0x0',
+      gasPrice: '0x0',
+    },
+  },
+  recipient: {
+    // Defines which mode to use for searching for matches in the input field
+    mode: RECIPIENT_SEARCH_MODES.CONTACT_LIST,
+    // Partial, not yet validated, entry into the address field. Used to share
+    // user input amongst the AddRecipient and EnsInput components.
+    userInput: '',
+    // The address of the recipient
+    address: '',
+    // The nickname stored in the user's address book for the recipient address
+    nickname: '',
+    // Error to display on the address field
+    error: null,
+  },
+};
+
+const slice = createSlice({
+  name,
+  initialState,
+  reducers: {
+    /**
+     * update current amount.value in state and run post update validation of
+     * the amount field and the send state. Recomputes the draftTransaction
+     */
+    updateSendAmount: (state, action) => {
+      state.amount.value = addHexPrefix(action.payload);
+      // Once amount has changed, validate the field
+      slice.caseReducers.validateAmountField(state);
+      // validate send state
+      slice.caseReducers.validateSendState(state);
+      // Recompute the draftTransaction object
+      slice.caseReducers.updateDraftTransaction(state);
+    },
+    /**
+     * computes the maximum amount of asset that can be sent and then calls
+     * the updateSendAmount action above with the computed value, which will
+     * revalidate the field and form and recomputes the draftTransaction
+     */
+    updateAmountToMax: (state) => {
+      let amount = '0x0';
+      if (state.asset.type === ASSET_TYPES.TOKEN) {
+        const decimals = state.asset.details?.decimals ?? 0;
+        const multiplier = Math.pow(10, Number(decimals));
+
+        amount = multiplyCurrencies(state.asset.balance, multiplier, {
+          toNumericBase: 'hex',
+          multiplicandBase: 16,
+          multiplierBase: 10,
+        });
+      } else {
+        amount = subtractCurrencies(
+          addHexPrefix(state.asset.balance),
+          addHexPrefix(state.gas.gasTotal),
+          {
+            toNumericBase: 'hex',
+            aBase: 16,
+            bBase: 16,
+          },
+        );
+      }
+      slice.caseReducers.updateSendAmount(state, {
+        payload: amount,
+      });
+      // draftTransaction update happens in updateSendAmount
+    },
+    /**
+     * updates the draftTransaction.userInputHexData state key and then
+     * recomputes the draftTransaction if the user is currently sending the
+     * native asset. When sending ERC20 assets, this is unnecessary because the
+     * hex data used in the transaction will be that for interacting with the
+     * ERC20 contract
+     */
+    updateUserInputHexData: (state, action) => {
+      state.draftTransaction.userInputHexData = action.payload;
+      if (state.asset.type === ASSET_TYPES.NATIVE) {
+        slice.caseReducers.updateDraftTransaction(state);
```

Comment:
Why is the draft tx only being updated with user input hex if the tx is a native type? On first read this looks like it would interfere with a user's ability to add custom hex data to a tx that is going out to a contract, which is when it's most meaningful.

Reply:
That's the current behavior and this doesn't aim to fix any current issues

Reply:
https://github.com/MetaMask/metamask-extension/issues/10194